### ISSUE TITLE
Some edge directions are opposite to the tutrial results. (version 0.5.1)

### ISSUE THREAD

**Issue by mn-tk** [erdogant/bnlearn#46]
Currently, when I ran bnlearn.ipynb in Colab with bnlearn==0.5.1 and pgmpy==0.1.17 (https://colab.research.google.com/github/erdogant/bnlearn/blob/master/notebooks/bnlearn.ipynb)
The edge of "Sprinkler"-"Cloudy" and the edge of "Cloudy"-"Rain" are opposite to the tutrial results.

![image](https://user-images.githubusercontent.com/24709629/151122204-d1572036-4a2a-48ec-81f7-6f96d2f6ad06.png)

Forthermore I ran bnlearn.ipynb in Colab with bnlearn==0.4.11 and pgmpy==0.1.17 as well, and the results were the same as the tutrial results.

![image](https://user-images.githubusercontent.com/24709629/151122262-9976fc79-746b-4826-8ac3-68b7f771b2d8.png)

Why do the results change due to the upgrade of bnlearn?

**Comment by erdogant** [comment]
Thanks for looking carefully! I could not find any obvious code differences in the bnlearn code that could result in the difference. All the input parameters are the same, and there are also no differences in the example data. I did however see an differences deeper in the code of pgmpy.  In the hillclimbsearch.py, line 318, the edges are differently created for some reason.

I stripped down the entire code to these few lines of pgmpy.

```
import bnlearn as bn
print(bn.__version__)

from pgmpy.estimators import BicScore
from pgmpy.estimators import HillClimbSearch
df = bn.import_example('sprinkler')

scoring_method = BicScore(df)
model = HillClimbSearch(df)
best_model = model.estimate(scoring_method=scoring_method, tabu_length=100, epsilon=0.0001, max_iter=1000000, fixed_edges=set(), show_progress=False)
print(best_model.edges())

```

Thus Im not sure why the differences occur, I only see that they seem to happened deeper in the code of pgmpy.

**Comment by erdogant** [comment]
I fixed the issue. Can you try again the [colab](https://colab.research.google.com/github/erdogant/bnlearn/blob/master/notebooks/bnlearn.ipynb)

**Comment by mn-tk** [comment]
Thank you for your fast fix
When I try again with Colab, I seem to get different results for the edges
again.
[image: image.png]
but I it works in the local conda environment on a Mac.
It may not behave well in a Colab environment.
------------------------------------------------------------------------------
Takatsugu KOSUGI


2022年1月28日(金) 20:53 Erdogan Taskesen ***@***.***>:

**Comment by mn-tk** [comment]
I was wrong, but I wasn't able to reproduce the edge direction （Sprinkler
-> Cloundy）when I did it in my local environment.
It could be the effect of pgmpy as you say.
but it's a bit dangerous that the direction of the edge changes depending
on the execution environment.
[image: image.png]
------------------------------------------------------------------------------
Takatsugu KOSUGI


2022年1月31日(月) 10:25 小杉孝嗣 ***@***.***>:

**Comment by erdogant** [comment]
It took a while to figure this out. 

The HillClimbSearch iteratively makes an operation on the DAG and updates the score. First, a **Set** is created from **potential_new_edges**  in **_legal_operations()** and is then returned together with the **score_delta**. The following step is that the function **estimate()** computes the **score_delta** to determine the **best_operation**.

Here comes the reason for your issue; (1) The  **_legal_operations()** uses a **Set** for which the order is not pre-defined and also changes for each Python run (check this note https://docs.python.org/3/reference/datamodel.html#object.__hash__), and (2) Multiple edges exists with exactly the same (maximum) **score_delta**. Because of points 1 and 2 together, each time you run the code, different operations can be used to iteratively build the DAG, and thus with different outcomes. 

If the sprinkler example would have contained a larger DAG, (more nodes or states), then HillClimbSearch may not result in exactly the same maximum **score_delta** values for different edges. However, this is still one of the 2 points that create the difference. 

The bottom line is; this issue is not a bug but the result of finding exactly the same scores for edges together with Python functionality that handles the ordering. This would also explain that different systems, notebooks, versions, or simply runs can result in differences.

**Comment by erdogant** [comment]
Dear Takatsugu,

This was really a difficult issue to figure out! But I found it.
I did also post it on: https://github.com/erdogant/bnlearn/issues/46

The HillClimbSearch iteratively makes an operation on the DAG and updates
the score. First, a Set is created from potential_new_edges in
_legal_operations() and is then returned together with the score_delta. The
following step is that the function estimate() computes the score_delta to
determine the best_operation.


Here comes the reason for your issue; (1) The _legal_operations() uses
a Set for
which the order is not pre-defined and also changes for each Python run
(check this note
https://docs.python.org/3/reference/datamodel.html#object.__hash__), and
(2) Multiple edges exists with exactly the same (maximum) score_delta.
Because of points 1 and 2 together, each time you run the code, different
operations can be used to iteratively build the DAG, and thus with
different outcomes.

If the sprinkler example would have contained a larger DAG, (more nodes or
states), then HillClimbSearch may not result in exactly the same maximum
score_delta values for different edges. However, this is still one of the 2
points that create the difference.

The bottom line is; this issue is not a bug. It simply finds exactly the
same scores for edges together with Python functionality that handles the
ordering. This would also explain that different systems, notebooks,
versions, or simply runs can result in differences.


I hope this explanation helps you.


Best,

Erdogan



Op ma 31 jan. 2022 om 08:16 schreef Takatsugu KOSUGI <
***@***.***>:

**Comment by erdogant** [comment]
If there are any other questions regarding this issue, let me know.
